### PR TITLE
Add linear layer in model for numerical gradient tests

### PR DESCRIPTION
**Pull request by saksham189**
I have added a linear layer for all numerical gradient tests behind the layer being tested.
This PR also include a fix in the implementation of `Backward` for BatchNorm layer. 

Closes https://github.com/mlpack/mlpack/issues/1923

### CODE REVIEW

**Comment by saksham189** [comment]
The `LayerNorm` gradient test was failing. I added the same fix as for `BatchNorm` and it seems to pass after that.

**Comment by saksham189** [comment]
Seems all the numerical gradient tests are passing after adding the linear layer.
@ShikharJ  let me know your thoughts.

**Comment by ShikharJ** [comment]
> The `LayerNorm` gradient test was failing. I added the same fix as for `BatchNorm` and it seems to pass after that.

That's probably because `LayerNorm` directly borrows from `BatchNorm`.

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 1]
```diff
@@ -25,7 +25,8 @@ BatchNorm<InputDataType, OutputDataType>::BatchNorm() :
     eps(1e-8),
     loading(false),
     deterministic(false),
-    count(0)
+    count(0),
+    size(0)
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp]: This should be introduced in the same order as the other constructor to prevent compile time warnings.

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 2]
```diff
@@ -80,16 +81,17 @@ void BatchNorm<InputDataType, OutputDataType>::Forward(
...
     for (size_t i = 0; i < input.n_cols; i++)
     {
       count += 1;
 
-      OutputDataType delta = input.col(i) - runningMean;
-      runningMean = runningMean + delta / count;
-      runningVariance += delta % (input.col(i) - runningMean);
+      OutputDataType diff = input.col(i) - runningMean;
+      runningMean = runningMean + diff / count;
+      runningVariance += diff % (input.col(i) - runningMean);
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp]: I think we should revert these changes, since they only change the name.

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp [hunk 3]
```diff
@@ -106,7 +108,6 @@ template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Backward(
     const arma::Mat<eT>&& input, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
 {
-  const arma::mat inputMean = input.each_col() - mean;
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/batch_norm_impl.hpp]: I see, we were not using the processed output. I remember facing such an issue in a part implementation of `ELU` as well. Nice catch.

File: src/mlpack/methods/ann/layer/layer_norm.hpp [hunk 4]
```diff
@@ -184,6 +184,9 @@ class LayerNorm
 
   //! Locally-stored normalized input.
   OutputDataType normalized;
+
+  //! Locally-stored input with 0 mean.
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/layer_norm.hpp]: Same here.

File: src/mlpack/methods/ann/layer/layer_norm_impl.hpp [hunk 5]
```diff
@@ -23,7 +23,8 @@ namespace ann { /** Artificial Neural Network. */
 template<typename InputDataType, typename OutputDataType>
 LayerNorm<InputDataType, OutputDataType>::LayerNorm() :
     eps(1e-8),
-    loading(false)
+    loading(false),
+    size(0)
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/layer_norm_impl.hpp]: Same here.

File: src/mlpack/methods/ann/layer/batch_norm_impl.hpp (hunk 2 above)
Comment (saksham189) [reply src/mlpack/methods/ann/layer/batch_norm_impl.hpp]: I did this change because `delta` is also a class member and hence it is shadowing the class variable. IMO it is best to keep the names separate to avoid mistakes when reading and modifying the code.

**Comment by ShikharJ** [comment]
Thanks @saksham189 for the fix.
